### ISSUE TITLE
Stream Timeout only seems to time out once and only after an item is emitted

### ISSUE THREAD

**Issue by Luminoth** [tokio-rs/tokio#2566]
**Version**
0.2.21

**Platform**
64-bit Windows 10

**Description**

I'm using tokio::stream::StreamExt::timeout() on the Stream returned by redis::aio::PubSub::on_message() in the redis crate. My timeout has a duration of 1 second. I'm seeing, however, timeouts coming only after a non-timeout occurs and only once. This is, I think, the relevant code snippet:

```
tokio::spawn(async move {
[...]
let mut msgs = pubsub.on_message().timeout(Duration::from_secs(1));

while !notifier.load(Ordering::SeqCst) {
    match msgs.try_next().await {
        Ok(msg) => match msg {
            Some(msg) => {
                info!("pubsub message");
            }
            None => {
                error!("pubsub disconnect");
                break;
            }
        },
        Err(_timeout) => {
            error!("pubsub timeout");
        }
    }
}
[...]
})
```

And this is a sample of the logs:

```
2020-05-28T00:59:09 "pubsub message"
2020-05-28T00:59:10 "pubsub timeout"
2020-05-28T00:59:14 "pubsub message"
2020-05-28T00:59:15 "pubsub timeout"
```

The notifier there is just an Arc<AtomicBool> that another task can set to tell that task to break out and exit. As far as I can tell, I'm not starving the executor anywhere, I see output from other tasks that are also running on it. The fact that the stream does continue to emit items also makes me think that it's not being starved but rather something might be wrong with the StreamExt::timeout() method.

**Comment by Luminoth** [comment]
Looking at the code for the stream timeout I think this is now making sense. The poll_deadline var is only being set true after the stream emits an item. I'm not certain if that's actually the intended behavior, but if so then I'm not sure how to refactor what I'm doing here so that my task can exit when it's told to.

**Comment by Darksonn** [comment]
How about wrapping the call to `StreamExt::next` in [`timeout`][1]? As for exiting on a signal, consider using a [`broadcast`][2] channel to signal shutdown, which you can combine with [`select!`][3] to exit immediately when the signal is sent.

[1]: https://docs.rs/tokio/0.2/tokio/time/fn.timeout.html
[2]: https://docs.rs/tokio/0.2/tokio/sync/broadcast/index.html
[3]: https://docs.rs/tokio/0.2/tokio/macro.select.html

**Comment by Luminoth** [comment]
That's fantastic, I tried the first suggestion and it worked perfectly. Thank you!
